mesonbuild/mcompile.py

```diff
@@ -0,0 +1,136 @@
+# Copyright 2020 The Meson development team
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Entrypoint script for backend agnostic compile."""
+
+import os
+import pathlib
+import shutil
+import subprocess
+import sys
+import typing as T
+
+from . import mlog
+
+if T.TYPE_CHECKING:
+    import argparse
+
+
+def add_arguments(parser: 'argparse.ArgumentParser') -> None:
+    """Add compile specific arguments."""
+    parser.add_argument(
+        '-j', '--jobs',
+        action='store',
+        default=0,
+        type=int,
+        help='The number of worker jobs to run (if supported). If the value is less than 1 the build program will guess.'
+    )
+    parser.add_argument(
+        '-l', '--load-average',
+        action='store',
+        default=0,
+        type=int,
+        help='The system load average to try to maintain (if supported)'
+    )
+    parser.add_argument(
+        '--debug',
+        action='store_true',
+        help='Print debug information for the meson compile command.'
+    )
+    parser.add_argument(
+        '--clean',
+        action='store_true',
+        help='Clean the build directory.'
+    )
+    parser.add_argument(
+        'builddir',
+        type=pathlib.Path,
+        nargs='?',
+        default='.',
+        help='The directory containing build files to be built.'
+    )
+
+
+def _run(options: 'argparse.Namespace') -> int:
+    if options.debug:
+        mlog.log_file = sys.stderr
+        mlog.set_verbose()
```

Comment:
Shouldn't we call mlog.initialize()?

Reply:
I don't think so, we're not using a logfile in the sense of a file, but putting everything on stderr, and then letting the subprocess call write to our stdout and stderr. If we initialize it'll try to open `sys.stderr`, or we'll end up writing files in the build directory? We only need this to get options of lower status than warning anyway.

Reply:
Hm, right, but since it's not logging into a real file, I'm not sure mlog.shutdown() will do. Any specific reason you want to print to stderr instead of stdout, btw? Ideally mlog.warning() should go to stderr but mlog.log() should go to stdout, I guess.

Also, verbose seems to be the default, I think you should call mlog.set_quiet() in else clause.

Reply:
I don't think we need this at all, actually if we just let mesonmain handle it.